File: DetectELJ/Python/WIP2/BurpExtenderWIP2.py
#
# DetectELJ - Active scanner extension to detect Expression Language Injection vulnerabilities
#
# Copyright (c) 2017 Doyensec LLC. Made with love by Andrea Brancaleoni.
#
from burp import IBurpExtender, IScannerCheck, IScanIssue
from java.io import PrintWriter
from java.net import URL
from java.util import ArrayList, Arrays, List

from org.python.core.util import StringUtil
from jarray import array


class BurpExtender(IBurpExtender, IScannerCheck):

    def registerExtenderCallbacks(self, callbacks):
        self.callbacks = callbacks
        self.helpers = callbacks.getHelpers()
        callbacks.setExtensionName('DetectELJ')

        callbacks.issueAlert('DetectELJ Active Scanner check enabled')

        stdout = PrintWriter(callbacks.getStdout(), True)
        stderr = PrintWriter(callbacks.getStderr(), True)

        callbacks.registerScannerCheck(self)

    def doPassiveScan(self, ihrr):
        return None # Active scanner check only

    def doActiveScan(self, ihrr, isip):
        # 1 - Create a new request with our custom payload (tip: buildRequest)
        withPayload = isip.buildRequest(StringUtil.toBytes('${1336+1}'))
        # 2 - Send the HTTP request
        newReqRes = self.callbacks.makeHttpRequest(ihrr.getHttpService(), withPayload)
        # 3 - Diff original and new responses (tip: analyzeResponseVariations and getVariantAttributes)
        variation = self.helpers.analyzeResponseVariations(ihrr.getResponse(), newReqRes.getResponse())
        pageChanges = variation.getVariantAttributes()
        # 4 - Based on page changes, determine whether the page is vulnerable or not
        length = False
        bodyContent = False
        match = False

        for change in pageChanges:
            if change == 'content_length':
                length = True
            if change == 'whole_body_content':
                bodyContent = True

        match = '1337' in self.helpers.bytesToString(newReqRes.getResponse())
        # 5 - If vulnerable, create a new IScanIssue and return the List<IScanIssue>
        if length and bodyContent and match:
            pass
            # TODO
        else:
            return None

    def consolidateDuplicateIssues(self, isb, isa):
        # TODO
        # If it is the same URL and same type of response, consider as the same security issue, otherwise different
        pass


class ELJ(IScanIssue):
    # TODO
    pass

    def __init__(self, reqres, callbacks, helpers):
        self.reqres = reqres
        self.callbacks = callbacks
        self.helpers = helpers

    def getHost(self):
        return self.reqres.getHost()

    def getPort(self):
        return self.reqres.getPort()

    def getProtocol(self):
        return self.reqres.getProtocol()

    def getUrl(self):
        return self.reqres.getUrl()

    def getIssueName(self):
        return 'Expression Language (EL) Injection Detected'

    def getIssueType(self):
        return 0x08000000  # See http://portswigger.net/burp/help/scanner_issuetypes.html

    def getSeverity(self):
        return 'High'  # 'High', 'Medium', 'Low', 'Information' or 'False positive'

    def getConfidence(self):
        return 'Firm'  # 'Certain', 'Firm' or 'Tentative'

    def getIssueBackground(self):
        return str('Expression Language injections occur when input data is evaluated '
                   'by an expression language interpreter. An attacker can read server-side '
                   'data, such as the content of server-side variables, and some other inner '
                   'configuration details.')

    def getRemediationBackground(self):
        return 'Apply input validation best practices, and reject $, # and other variations.'

    def getIssueDetail(self):
        return str('Burp Scanner has identified an Expression Language injection in:<b>'
                   '%s</b><br><br>' % (self.reqres.getUrl().toString()))

    def getRemediationDetail(self):
        return None

    def getHttpMessages(self):
        # Let's highlight the specific string in the response that triggered the issue

        # TODO
        pass

    def getHttpService(self):
        return self.reqres.getHttpService()

File: DetectELJ/Ruby/WIP2/BurpExtenderWIP2.rb
#
# DetectELJ - Active scanner extension to detect Expression Language Injection vulnerabilities
#
# Copyright (c) 2017 Doyensec LLC. Made with love by Andrea Brancaleoni.
#
require 'java'

java_import 'burp.IBurpExtender'
java_import 'burp.IScannerCheck'
java_import 'burp.IScanIssue'
java_import 'burp.IHttpRequestResponse'
java_import 'java.io.PrintWriter'
java_import 'java.net.URL'
java_import 'java.util.ArrayList'
java_import 'java.util.Arrays'
java_import 'java.util.List'

class BurpExtender
  include IBurpExtender
  include IScannerCheck

  attr_reader :helpers
  attr_reader :callbacks

  def registerExtenderCallbacks(callbacks)
    @callbacks = callbacks
    @helpers = callbacks.getHelpers()
    callbacks.setExtensionName('DetectELJ')

    callbacks.issueAlert('DetectELJ Active Scanner check enabled')

    stdout = PrintWriter.new(callbacks.getStdout(), true)
    stderr = PrintWriter.new(callbacks.getStderr(), true)

    callbacks.registerScannerCheck(self)
  end

  def print(e)
    @callbacks.issueAlert(e.to_s)
  end

  def doPassiveScan(ihrr)
    return nil # Active scanner check only
  end


  def doActiveScan(ihrr, isip)
    # 1 - Create a new request with our custom payload (tip: buildRequest)
    withPayload = isip.buildRequest('${1336+1}'.to_java_bytes())
    # 2 - Send the HTTP request
    newReqRes = callbacks.makeHttpRequest(ihrr.getHttpService(), withPayload)
    # 3 - Diff original and new responses (tip: analyzeResponseVariations and getVariantAttributes)
    variation = helpers.analyzeResponseVariations(ihrr.getResponse(), newReqRes.getResponse())
    pageChanges = variation.getVariantAttributes()
    # 4 - Based on page changes, determine whether the page is vulnerable or not
    length = false
    bodyContent = false
    match = false

    pageChanges.each do |change|
      length = true if change == 'content_length'
      bodyContent = true if change == 'whole_body_content'
    end

    # XXX: ruby string contains shim
    match = helpers.bytesToString(newReqRes.getResponse())['1337']
    # 5 - If vulnerable, create a new IScanIssue and return the List<IScanIssue>
    if (length && bodyContent && match)
      # TODO
    else
      return nil
    end
  end


  def consolidateDuplicateIssues(isb, isa)
    # TODO
  end

end

class ELJ
  include IScanIssue
  attr_reader :helpers
  attr_reader :reqres
  attr_reader :callbacks

  def initialize(reqres, callbacks, helpers)
    @reqres = reqres
    @callbacks = callbacks
    @helpers = helpers
  end


  def getHost()
    return reqres.getHost()
  end


  def getPort()
    return reqres.getPort()
  end


  def getProtocol()
    return reqres.getProtocol()
  end


  def getUrl()
    return reqres.getUrl()
  end


  def getIssueName()
    return 'Expression Language (EL) Injection Detected'
  end


  def getIssueType()
    return 0x08000000 # See http://portswigger.net/burp/help/scanner_issuetypes.html
  end


  def getSeverity()
    return 'High' #  'High', 'Medium', 'Low', 'Information' or 'False positive'
  end


  def getConfidence()
    return 'Firm' # 'Certain', 'Firm' or 'Tentative'
  end


  def getIssueBackground()
    return 'Expression Language injections occur when input data is evaluated ' +
    'by an expression language interpreter. An attacker can read server-side ' +
    'data, such as the content of server-side variables, and some other inner ' +
    'configuration details.'
  end


  def getRemediationBackground()
    return 'Apply input validation best practices, and reject $, # and other variations.'
  end


  def getIssueDetail()
    return 'Burp Scanner has identified an Expression Language injection in:<b>' +
            reqres.getUrl().toString() + '</b><br><br>'
  end


  def getRemediationDetail()
    return nil
  end


  def getHttpMessages()
    # Let's highlight the specific string in the response that triggered the issue
    # TODO
  end


  def getHttpService()
    return reqres.getHttpService()
  end
end
